### ISSUE TITLE
CanvasMarker - HowTo

### ISSUE THREAD

**Issue by emVdovin** [infusion-code/angular-maps#2]
Hi 

Do you have an example on how to use CanvasMarker? Can you post it? It will be very helpful

Regards,
Evgeny

**Comment by thor-schueler** [comment]
Sorry, for the late response. Yes, I'll put on on plunker the next days or so... But it boils down to setting your icon info on the marker like so:

```
<x-map-marker                             
   [Latitude]="..."
   [Longitude]="..."
   [Title]="...."
   [IconInfo]="{
      id: "someid"
      markerType: _markerTypeId.CanvasMarker,
      size: { width: 100, height: 100 },
      points: [
         { x: x1, y: y1},
         { x: x2, y: y2},
         { x:  x3, y: y3},
         { x: x4, y: y4},
      ]
      markerOffsetRatio: { x: 0.5, y: 0.5 }           
   }"
   (MarkerClick)="...">
</x-map-marker>
```

with canvas based markers, you need to supply the desired size of your marker in pixels (type` ISize`) and the points to draw on the canvas (an array of `IPoints`). Both of these are part of the `IMarkerIconInfo` you attach to the `IconInfo` property of the marker (`IMarkerIconInfo.size` and `IMarkerIconInfo.points` respectively). 

The canvas icon will be drawn on the canvas by iterating over the `IPoints` in  `IMarkerIconInfo.points` and drawing a line from one point to the next. So the order of the points does matter. The start point of the path is determined by `IMarkerIconInfo.drawingOffset`. If that is not present, the path starts at `{0, 0}`.  You can also specify  `IMarkerIconInfo.rotation` if you want to rotate you marker on the canvas. Depending in where the sharp point of your marker is, you can use `IMarkerIconInfo.markerOffsetRatio` to shift the marker vertically or horizontally by a percentage of its size, so as to align the sharp point with the actual geo coordinate of the map. The might require some tweaking, but generally, you want to make sure you marker does not drift when you zoom in/out. 

Generally, when I do canvas icons, I start with some graph paper to draw out the icon i want to create. That will give me the succession of points.  

You can try it out with the following, which should give you a little red rotated arrow like marker:

```
const iconInfo: IMarkerIconInfo = {
     markerType: MarkerTypeId.CanvasMarker,
     rotation: 45,
     drawingOffset: { x: 24, y: 0 },
     points: [
       { x: 10, y: 40 },
       { x: 24, y: 30 },
       { x: 38, y: 40 },
       { x: 24, y: 0 }
     ],
     color: '#f00',
     size: { width: 48, height: 48 }
};
```

**Comment by emVdovin** [comment]
Thanks a lot. I actually figured this out and it works perfectly.
One thing I can not figure out is a scaledimagemarker.
I can not see my icon always getting issue with callback.
If you have a sample it will be appreciated

Regards

Evgeny Vdovin
Altyn Consulting Inc
m:      778-928-0829<tel:778-928-0829>  e: evgeny.vdovin@altyn.ca<mailto:evgeny.vdovin@altyn.ca>
a:      1066 West Hastings, Suite 2300, Vancouver, BC, V6E 3X2<x-apple-data-detectors://1/3>

<http://www.hubspot.com/www.linkedin.com/in/AltynConsulting>

[cid:5898121F-FD9E-45C2-BFC3-23D49A821E4B]

Sent from my iPhone

On Jul 1, 2017, at 9:34 AM, Thor Schueler <notifications@github.com<mailto:notifications@github.com>> wrote:


Sorry, for the late response. Yes, I'll put on on plunker the next days or so... But it boils down to setting your icon info on the marker like so:

<x-map-marker
   [Latitude]="..."
   [Longitude]="..."
   [Title]="...."
   [IconInfo]="{
      id: "someid"
      markerType: _markerTypeId.CanvasMarker,
      size: { width: 100, height: 100 },
      points: [
         { x: x1, y: y1},
         { x: x2, y: y2},
         { x:  x3, y: y3},
         { x: x4, y: y4},
      ]
      markerOffsetRatio: { x: 0.5, y: 0.5 }
   }"
   (MarkerClick)="...">
</x-map-marker>


with canvas based markers, you need to supply the desired size of your marker in pixels (typeISize) and the points to draw on the canvas (an array of IPoints). Both of these are part of the IMarkerIconInfo you attach to the IconInfo property of the marker (IMarkerIconInfo.size and IMarkerIconInfo.points respectively).

The canvas icon will be drawn on the canvas by iterating over the IPoints in IMarkerIconInfo.points and drawing a line from one point to the next. So the order of the points does matter. The start point of the path is determined by IMarkerIconInfo.drawingOffset. If that is not present, the path starts at {0, 0}. You can also specify IMarkerIconInfo.rotation if you want to rotate you marker on the canvas. Depending in where the sharp point of your marker is, you can use IMarkerIconInfo.markerOffsetRatio to shift the marker vertically or horizontally by a percentage of its size, so as to align the sharp point with the actual geo coordinate of the map. The might require some tweaking, but generally, you want to make sure you marker does not drift when you zoom in/out.

Generally, when I do canvas icons, I start with some graph paper to draw out the icon i want to create. That will give me the succession of points.

You can try it out with the following, which should give you a little red rotated arrow like marker:

const iconInfo: IMarkerIconInfo = {
     markerType: MarkerTypeId.CanvasMarker,
     rotation: 45,
     drawingOffset: { x: 24, y: 0 },
     points: [
       { x: 10, y: 40 },
       { x: 24, y: 30 },
       { x: 38, y: 40 },
       { x: 24, y: 0 }
     ],
     color: '#f00',
     size: { width: 48, height: 48 }
};

**Comment by emVdovin** [comment]
Thanks for closing this issue but how about some new samples :)

We really need one that will show how to use a callback for scaledimagemarker and I guess it will be same for the rest

**Comment by thor-schueler** [comment]
Yes, will put one out there,  over the next few days. You are correct, all the callback based ones work the same way. The need for the callback is rooted in the fact that those methods use html elements to do the actual work (like an image element) which effectively work asynchronously. 

Generally, the principle is to pass a callback in `IMarkerIconInfo`. This call back is called once the actual marker creation is complete with two parameters: a string containing the icon data uri and the `IMarkerIconInfo` populated with the actual icon size (in `IMarkerIconInfo.size`). 

For markers created that way, they will be initially created with an empty pushpin (as those methods return an empty string). Therefore, in the callback the pushpin will need to be updated the the actual marker... I guess it'd be nicer if I would incorporate this natively in the CreateMarker method of the MapService....

**Comment by thor-schueler** [comment]
@emVdovin : re the scaledimagemarker... I was in the process of putting an example together, but doing so it occurred to me that this is unnecessarily convoluted (all the callback markers really). So I refactored to make the callback execution internal. As a result, the callback markers will work the same way as the other markers. You can checkout the changes in the callback-markers branch. I validated the functionality with Bing, but need to do some more testing in Google. Also, there are some pending pull requests I need to process first. I think I'll have this stuff merged into the main branch by the end of the week and published.

**Comment by emVdovin** [comment]
Great news.
 Thanks I will test it by end of the week

Regards

Evgeny Vdovin
Altyn Consulting Inc
m:      778-928-0829<tel:778-928-0829>  e: evgeny.vdovin@altyn.ca<mailto:evgeny.vdovin@altyn.ca>
a:      1066 West Hastings, Suite 2300, Vancouver, BC, V6E 3X2<x-apple-data-detectors://1/3>

<http://www.hubspot.com/www.linkedin.com/in/AltynConsulting>

[cid:5898121F-FD9E-45C2-BFC3-23D49A821E4B]

Sent from my iPhone

On Jul 25, 2017, at 4:04 PM, Thor Schueler <notifications@github.com<mailto:notifications@github.com>> wrote:


@emVdovin<https://github.com/emvdovin> : re the scaledimagemarker... I was in the process of putting an example together, but doing so it occurred to me that this is unnecessarily convoluted (all the callback markers really). So I refactored to make the callback execution internal. As a result, the callback markers will work the same way as the other markers. You can checkout the changes in the callback-markers branch. I validated the functionality with Bing, but need to do some more testing in Google. Also, there are some pending pull requests I need to process first. I think I'll have this stuff merged into the main branch by the end of the week and published.

**Comment by thor-schueler** [comment]
@emVdovin: I merged the marker changes and published in version 0.5.4, I added some samples for markers as well. Please see the sample page in the wiki: https://github.com/infusion-code/angular-maps/wiki/Samples.
